docs/cn/built_in_function_develop_guide.md

```diff
@@ -0,0 +1,230 @@
+
+```c++
+namespace v1 {
+  void timestamp_to_date(codec::Timestamp *timestamp,
+       									 codec::Date *output /*result output*/, bool *is_null /*null flag*/) {
+    time_t time = (timestamp->ts_ + TZ_OFFSET) / 1000;
+    struct tm t;
+    if (nullptr == gmtime_r(&time, &t)) {
+        *is_null = true;
+        return;
+    }
+    *output = codec::Date(t.tm_year + 1900, t.tm_mon + 1, t.tm_mday);
+    *is_null = false;
+    return;
+	}
+} // namespace v1
+```
+
+#### step 2: 配置参数，返回值并注册函数
+
+**示例:**
+
+函数名和函数参数的配置和普通函数配置一样。但需要额外注意返回值类型的配置：
+
+- 因为函数结果存放在参数中返回，所以配置`return_by_arg(true)`
+- 因为函数结果可能为null,所以配置`.returns<Nullable<Date>>`
+
+```c++
+RegisterExternal("date")
+        .args<codec::Timestamp>(reinterpret_cast<void*>(
+            static_cast<void (*)(Timestamp*, Date*, bool*)>(
+                v1::timestamp_to_date)))
+        .return_by_arg(true)
+        .returns<Nullable<Date>>()
+        .doc(R"(
+            @brief Cast timestamp or string expression to date
+
+            Example:
+
+            @code{.sql}
+                select date(timestamp(1590115420000));
+                -- output 2020-05-22
+                select date("2020-05-22");
+                -- output 2020-05-22
+            @endcode
+            @since 0.1.0)");
+```
+
+## 函数文档配置
+
+`ExternalFuncRegistryHelper` 提供了`doc(doc_string)`API来配置文档。函数文档用来描述函数的功能和用法。它的主要受众是普通用户。所以，文档需要清晰易懂。
+
+函数文档一般需要包含一下几类信息：
+
+- **@brief** ：简要描述函数的功能
+- **@param**：函数参数描述
+- **Examples**： 函数的使用示例。一般需要提供SQL的查询示例。SQL语句需要放置在 a `@code/@endcode` 代码块中。
+- **@since** ：指明引入该函数的OpenMLDB版本
+
+**示例:**
+
+```c++
+RegisterExternal("my_function")
+  			.args<int64_t>(static_cast<int32_t (*)(int64_t)>(v1::my_func))
+        .doc(R"(
+            @brief a brief summary of the my_function's purpose and behavior
+            
+            @param param1 a brief description of param1
+            
+            Example:
+
+            @code{.sql}
+                select my_function(1);
+                -- output xxx
+            @endcode
+            @since 0.4.0)");
+```
+
+## 函数测试
+
+函数开发完成后，开发者还需要添加相对应的测试例以确保系统运行正确。
+
+### `UdfIRBuilderTest`中添加单测
+
+一般地，可以在`src/codegen/udf_ir_builder_test.cc`中添加`TEST_F`单测。我们提供了CheckUdf函数以便开发者检验函数结果。
+
+```c++
+CheckUdf<return_type, arg_type,...>("function_name", expect_result, arg_value,...);
+```
+
+对每一个函数（每一种参数组合），我们至少需要实现:
+
+- 添加一个单测，验证普通结果
+- 如果结果是***nullable***,需要添加一个单测，验证结果为空
+
+**示例:**
+
+```c++
+// month(timestamp) normal check
+TEST_F(UdfIRBuilderTest, month_timestamp_udf_test) {
+    Timestamp time(1589958000000L);
+    CheckUdf<int32_t, Timestamp>("month", 5, time);
+}
+// date(timestamp) normal check
+TEST_F(UdfIRBuilderTest, timestamp_to_date_test_0) {
+    CheckUdf<Nullable<Date>, Nullable<Timestamp>>(
+        "date", codec::Date(2020, 05, 20), codec::Timestamp(1589958000000L));
+}
+// date(timestamp) null check
+TEST_F(UdfIRBuilderTest, timestamp_to_date_test_null_0) {
+    CheckUdf<Nullable<Date>, Nullable<Timestamp>>("date", nullptr, nullptr);
+}
```

Comment:
How to test end2end with sql cmd, link to the compile doc?

Reply:
I will add a section for:
# Integration Test Case